projects/briey/libs/ascon.c

```diff
@@ -0,0 +1,87 @@
+#include "ascon.h"
+
+#include <stdint.h>
+#include <string.h>
+
+static volatile struct {
+	uint32_t unique_value;
+	uint32_t status;
+	uint32_t control;
+	uint32_t scheduled_status;
+	uint32_t key[4];
+	uint32_t nonce[4];
+	uint32_t data[2];
+	uint32_t filler2[2];
+	uint32_t tag[4];
+} *const ASCON128_INTERFACE = (volatile void *) 0x40000000;
+
+#define ASCON_CTRL_INIT (1 << 0)
+#define ASCON_CTRL_ASSOCIATE (1 << 1)
+#define ASCON_CTRL_ENCRYPT (1 << 2)
+#define ASCON_CTRL_DECRYPT (1 << 3)
+#define ASCON_CTRL_ENCRYPT_FINAL (1 << 4)
+#define ASCON_CTRL_DECRYPT_FINAL (1 << 5)
+#define ASCON_CTRL_ASSOCIATE_FINAL (1 << 6)
+
+#define UNIQUE_VALUE 0xdeadbeef
+
+int ascon_is_sane(void)
+{
+	return ASCON128_INTERFACE->unique_value == UNIQUE_VALUE;
+}
+
+int ascon_is_busy(void)
+{
+	return ASCON128_INTERFACE->status;
+}
+
+static inline void ascon_busy_wait(void)
+{
+	while (ascon_is_busy())
+		;
+}
+
+static void ascon_memcpy(volatile uint32_t *dest, const volatile uint32_t *src, size_t size)
```

Comment:
Add restrict to the pointer qualifiers since we shouldn't be mem copying overlapping memory.